lib/discourse_redis.rb

```diff
@@ -4,139 +4,251 @@
+      @execution = execution
+      @master = true
+      @event_handlers = []
     end
 
-    def master
-      synchronize { @master }
+    def add_callbacks(handler)
+      @mutex.synchronize do
+        @event_handlers << handler
+      end
     end
 
-    def master=(args)
-      synchronize do
-        @master = args
+    def use_master?
+      master = @mutex.synchronize { @master }
+      if !master
+        return false
+      elsif safe_master_alive?
+        return true
+      else
+        was_reset =
+          @mutex.synchronize do
+            if @master
+              @master = false
+              trigger(:down)
+              true
+            else
+              false
+            end
+          end
+
+        if was_reset
+          @execution.spawn do
+            loop do
+              @execution.sleep 5
+              info "Checking connection to master"
+              if safe_master_alive?
+                @mutex.synchronize do
+                  @master = true
+                  @redis_status.fallback
+                  trigger(:up)
+                end
+                break
+              end
+            end
+          end
+        end
 
-        # Disables MessageBus keepalive when Redis is in readonly mode
-        MessageBus.keepalive_interval = 0 if !@master
+        return false
       end
     end
 
     private
 
-    def synchronize
-      @mutex.synchronize { yield }
+    attr_reader :log_prefix
+
+    def trigger(event)
+      @event_handlers.each do |handler|
+        begin
+          handler.public_send(event)
+        rescue Exception => e
+          Discourse.warn_exception(e, message: "Error running FallbackHandler callback")
+        end
+      end
     end
 
-    def logger
-      Rails.logger
+    def info(message)
+      Rails.logger.info "#{log_prefix}: #{message}"
     end
 
-    def log_prefix
-      "#{self.class}"
+    def safe_master_alive?
+      begin
+        @redis_status.master_alive?
+      rescue Exception => e
+        Discourse.warn_exception(e, message: "Error running master_alive?")
+        false
+      end
     end
   end
 
-  class Connector < Redis::Client::Connector
-    def initialize(options)
-      super(options)
-      @slave_options = DiscourseRedis.slave_config(options)
-      @fallback_handler = DiscourseRedis::FallbackHandler.instance
+  class MessageBusFallbackCallbacks
+    def down
+      @keepalive_interval, MessageBus.keepalive_interval =
+        MessageBus.keepalive_interval, 0
+    end
+
+    def up
+      MessageBus.keepalive_interval = @keepalive_interval
+    end
+  end
+
+  class MainRedisReadOnlyCallbacks
+    def down
+    end
+
+    def up
+      Discourse.clear_readonly!
+      Discourse.request_refresh!
+    end
+  end
+
+  class FallbackHandlers
+    include Singleton
+
+    def initialize
+      @mutex = Mutex.new
+      @fallback_handlers = {}
+    end
+
+    # Used by tests
+    def reset
+      @fallback_handlers = {}
```

Comment:
I think that you are using that reset method only in one place (in before block) do maybe instead of having that code here, we can use instance_variable_set in spec itself? https://apidock.com/ruby/Object/instance_variable_set

It is uglier, but on the other hand we will not put the code required by specs into the application.